File: src/mbs/domain/services/comment.py
import uuid
import datetime

from mbs.domain.models import Comment, CommentState, UserId, ArticleId, CommentId


class CommentService:

    def create_comment(self, content: str, author_id: UserId, article_id: ArticleId, parent_id: CommentId | None):
        return Comment(
            id=uuid.uuid4(),
            content=content,
            author_id=author_id,
            article_id=article_id,
            parent_id=parent_id,
            state=CommentState.PUBLISHED,
            created_at=datetime.datetime.now(datetime.UTC),
            updated_at=None
        )

    def update_comment(
            self,
            comment: Comment,
            content: str,
            state: CommentState,
    ):
        return Comment(
            id=comment.id,
            content=content,
            author_id=comment.author_id,
            state=state,
            created_at=comment.created_at,
            updated_at=datetime.datetime.now(datetime.UTC)
        )

File: src/domain/models/comment.rs
use chrono::{DateTime, Utc};
use crate::domain::models::{
    article::ArticleId,
    comment_state::CommentState,
    user_id::UserId
};

pub type CommentId = uuid::Uuid;
pub const COMMENT_CONTENT_MAX: usize = 1000;

pub struct Comment {
    pub id: CommentId,
    pub content: String,
    pub author_id: UserId,
    pub article_id: ArticleId,
    pub parent_id: Option<CommentId>,
    pub rating: i64,
    pub state: CommentState,

    pub created_at: DateTime<Utc>,
    pub updated_at: Option<DateTime<Utc>>,
}

impl Comment {
    pub fn new(content: String, author_id: UserId, article_id: ArticleId, parent_id: Option<CommentId>) -> Self {
        Self {
            id: CommentId::new_v4(),
            content,
            author_id,
            article_id,
            parent_id,
            rating: 0,
            state: CommentState::Published,
            created_at: Utc::now(),
            updated_at: None,
        }
    }

    pub fn update(&mut self, new_content: String) {
        self.content = new_content;
        self.updated_at = Some(Utc::now());
    }
    
    pub fn mark_deleted(&mut self) {
        self.state = CommentState::Deleted;
        self.updated_at = Some(Utc::now());
    }
}